### PR TITLE
Add npm publish command

### PR DESCRIPTION
**Pull request by guicaulada**

### CODE REVIEW

**Comment by guicaulada** [comment]
[Remove repeated implementations of ExtractedArchive](https://github.com/grafana/grafana-build/pull/115/commits/2d1dd61b03665412a534d0be70ad0e7b3c444c03)
[Add latest and main publishing to docker publish](https://github.com/grafana/grafana-build/pull/115/commits/9677d572289e386d20c9478b02aa42757211e29e)

Were a little bit out of scope, but seemed like a good time to fix this before I forget since both NPM and Docker have "latest" adjustments.

File: containers/npm_publish.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+	latestPreview, err := GetLatestGrafanaVersion(ctx, d, executil.Preview)
+	if err != nil {
+		return "", err
+	}
+
+	isLaterThanPreview, err := IsLaterThan(ctx, d, version, latestPreview)
+	if err != nil {
+		return "", err
+	}
```
Comment (kminehart) [inline containers/npm_publish.go:49]: Can you give me some context why all of this is necessary / what these functions do?

I'm a little bit worried that there's too much logic in this package that should maybe be delegated to the thing that is calling it, like for example including a `--latest` flag to publish the given npm package as the latest one, for example.

`grabpl` had a lot of similar logic and it made it really hard to follow and safely make changes to these different paths. We ended up overwriting things when we didn't mean to because a change looked benign that wasn't and so on.

Comment (kminehart) [reply containers/npm_publish.go:49]: A good way we can handle this, even would be to include an argument when promoting for "latest", for example, which could apply both to `npm` and `docker`.

**Review by kminehart — CHANGES_REQUESTED** [review]

File: containers/npm_publish.go (hunk 1 above)
Comment (guicaulada) [reply containers/npm_publish.go]: I removed this crazy version parsing logic and replaced it by the `latest` flag.

File: cmd/npm_publish.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"github.com/grafana/grafana-build/pipelines"
+	"github.com/urfave/cli/v2"
+)
+
+var PublishNPMCommand = &cli.Command{
+	Name:   "publish",
+	Action: PipelineActionWithPackageInput(pipelines.PublishNPM),
+	Usage:  "Using a grafana.tar.gz as input (ideally one built using the 'package' command), take the npm artifacts and publish them on NPM.",
```
Comment (kminehart) [inline cmd/npm_publish.go:11]: Is this description accurate for this command?

Comment (guicaulada) [reply cmd/npm_publish.go:11]: Yup! That's accurate, it takes a package built with the package command and publishes the npm-artifacts inside it.

File: containers/npm_publish.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+	major := strings.Split(version, ".")[0]
+	minor := strings.Split(version, ".")[1]
+	tag := fmt.Sprintf("latest-%s.%s", major, minor)
+
+	c := d.Container().From(NodeImage("lts")).
+		WithFile("/pkg.tgz", pkg).
+		WithExec([]string{"npm", "set", fmt.Sprintf("//%s/:_authToken", opts.Registry), opts.Token}).
```
[the remark this answers is not in the document]
Comment (guicaulada) [reply containers/npm_publish.go:31]: Nice catch :cold_sweat:

File: cmd/npm_publish.go (hunk 2 above)
Comment (guicaulada) [reply cmd/npm_publish.go:11]: This is to keep the .tar.gz built by the package command as the source of truth, instead of allowing any NPM package to be published by the command.

File: pipelines/docker_publish.go [hunk 4]
```diff
@@ -18,6 +18,17 @@ func ImageManifest(tag string) string {
...
+func LatestManifest(tag string, base BaseImage) string {
+	suffix := ""
+	if base == BaseImageUbuntu {
+		suffix = "-ubuntu"
+	}
+
+	manifest := strings.ReplaceAll(tag, "-image-tags", "")
+	manifestImage := strings.Split(manifest, ":")[0]
+	return strings.Join([]string{manifestImage, fmt.Sprintf("latest%s", suffix)}, ":")
+}
```
Comment (kminehart) [inline pipelines/docker_publish.go:30]: very testable function right here!

**Review by kminehart — APPROVED** [review]
Would like tos ee a test for the function I pointed out but otherwise it looks good to me.

File: pipelines/docker_publish.go (hunk 4 above)
Comment (guicaulada) [reply pipelines/docker_publish.go]: Done on 3b3312a
